plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3ExchangeStorage.java

```diff
@@ -0,0 +1,686 @@
+                bufferSize += copied;
+
+                try {
+                    flushBuffer(false);
+                }
+                catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+
+                offset += copied;
+                length -= copied;
+            }
+        }
+
+        @Override
+        public long getRetainedSize()
+        {
+            return sizeOf(buffer) + INSTANCE_SIZE;
+        }
+
+        @Override
+        public void close()
+                throws IOException
+        {
+            if (failed) {
+                try {
+                    abortUpload();
+                    return;
+                }
+                catch (RuntimeException e) {
+                    throw new IOException(e);
+                }
+            }
+
+            try {
+                flushBuffer(true);
+                waitForPreviousUploadFinish();
+            }
+            catch (IOException | RuntimeException e) {
+                abortUploadSuppressed(e);
+                throw e;
+            }
+
+            try {
+                uploadId.ifPresent(this::finishUpload);
+            }
+            catch (RuntimeException e) {
+                abortUploadSuppressed(e);
+                throw new IOException(e);
+            }
+        }
+
+        private void flushBuffer(boolean finished)
+                throws IOException
+        {
+            try {
+                waitForPreviousUploadFinish();
+            }
+            catch (IOException e) {
+                failed = true;
+                abortUploadSuppressed(e);
+                throw e;
+            }
+
+            // skip multipart upload if there would only be one part
+            if (finished && uploadId.isEmpty()) {
+                InputStream in = new ByteArrayInputStream(buffer, 0, bufferSize);
+
+                ObjectMetadata metadata = new ObjectMetadata();
+                metadata.setContentLength(bufferSize);
+                metadata.setContentMD5(getMd5AsBase64(buffer, 0, bufferSize));
+
+                PutObjectRequest request = new PutObjectRequest(bucketName, key, in, metadata);
+                sseCustomerKey.ifPresent(request::withSSECustomerKey);
+
+                try {
+                    s3.putObject(request);
+                    return;
+                }
+                catch (AmazonServiceException e) {
+                    throw new IOException(e);
+                }
+            }
+
+            // The multipart upload API only accept the last part to be less than 5MB
+            if (bufferSize == buffer.length || (finished && bufferSize > 0)) {
+                byte[] data = buffer;
+                int length = bufferSize;
+                this.buffer = new byte[buffer.length];
+                bufferSize = 0;
+
+                inProgressUploadFuture = uploadExecutor.submit(() -> uploadPage(data, length));
```

Comment:
could we use https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/s3/S3AsyncClient.html and not have executor on our side?
Just a question. Not sure if we are actually getting much/any benefit (depends on how S3AsyncClient is implemented)

Reply:
This is a good one. Let me do some research actually.